plugins/jenkins.js

```diff
@@ -1,20 +1,239 @@
 const _ = require('lodash');
-const xml = require('node-xml');
-const querystring = require('querystring');
 
 const core = require('../core.js');
 const ds = require('../datastore.js');
 const config = require('../configenv.js');
 
 var jenkinsStore = new ds.DataStore();
 
-/* * Given a job object, makes a build request to object's URL and
-   * communicates to user on given channel. Checks for failed builds
-   * and notifies user.
+/* *    Flow:
+   * executePlugin (determines which operation, parses given parameters) =>
+   * checkParams (checks that required params are provided) =>
+   * buildJenkinsJob (executes build with parameters, if specified) =>
+   * checkJobStatus (makes continual requests until job is dequeued) =>
+   * leftJobInfo & updateLeftJobStatus(periodically check status of running jobs)
+   *
+   *    Helper Functions:
+   * handleParameters - parses "KEY=val" strings into parameter objects
+   * handleListKeyword - returns list of Jenkins jobs with (optional) keyword
+   * findMatches - finds Jenkins jobs that match keyword
+   * getFullJobList - returns list of all Jenkins jobs
    */
-function buildJenkinsJob (requestedJobObject, channel, callback, parameters) {
 
-    console.log('PARAMETERS: ', parameters);
+
+function isCallable (text) {
+    return text.includes('jenkins');
+}
+
+
+function helpDescription () {
+    return '_JENKINS_\nSend *jenkins [keyword] list* to list jenkins' +
+    ' jobs with specified keyword in name.\n' +
+    'Send *jenkins [job name]* to build a jenkins job. If the job name is ' +
+    'not exactly correct, bot will attempt to fuzzy match it to the ' +
+    'correct job.\n Send *jenkins [job name] -p KEY=value* to build a job ' +
+    'with parameters';
+}
+
+
+function getFullJobList (callback) {
+    var options = {
+        url: 'jenkins.whoop.com/api/json',
+        method: 'POST'
+    };
+
+    core.makeRequest(options, function (dataObject) {
+        callback(dataObject.jobs);
+    });
+}
+
+
+function findMatches (keyword, collection, additionalFun) {
+    var regexp = new RegExp(keyword, 'i');
+    var foundMatches = [];
+    var counter = 0;
+
+    collection.forEach(function (item) {
+        if (regexp !== '' && regexp.test(item.name)) {
+            foundMatches.push(item);
+        }
+
+        if (additionalFun) {
+            additionalFun(item);
+        }
+
+        counter++;
+    });
+
+    if (counter === collection.length) {
+        return foundMatches;
+    }
+}
```

Comment:
I don't fully understand this function. If you're trying to check that every item satisfied the regex, you could try `_.every`

Do you mean to run `additionalFun` on every item, even if it doesn't match? You might do

```
collection.forEach(additionalFun);
```

If your list is small, it can be better to loop over it multiple times because the computational cost is negligible and the code is clearer. In this case, if running `additionalFun` is independent from checking the regex, then I think it's clearer to make it a separate statement.